cedar-policy-validator/src/validation_result.rs

```diff
@@ -84,13 +84,10 @@ pub struct ValidationError {
 }
 
 impl ValidationError {
-    pub(crate) fn with_policy_id(
-        id: PolicyID,
-        source_loc: Option<Loc>,
-        kind: ValidationErrorKind,
-    ) -> Self {
+    pub(crate) fn with_policy_id(id: PolicyID, source_loc: Option<Loc>, err: TypeError) -> Self {
+        let source_loc = err.source_loc().cloned().or(source_loc);
         Self {
-            kind,
+            kind: err.kind,
             location: SourceLocation::new(id, source_loc),
```

Comment:
I think there might be a `.kind_and_location()` to decompose `err` without cloning the source_loc?